tuta-sdk/rust/src/crypto/crypto_facade.rs

```diff
@@ -0,0 +1,401 @@
+        }
+
+        let owner_key_data = EntityOwnerKeyData::extract_owner_key_data(entity)?;
+
+        let EntityOwnerKeyData {
+            owner_enc_session_key: Some(owner_enc_session_key),
+            owner_key_version: Some(owner_key_version),
+            owner_group: Some(owner_group),
+            ..
+        } = owner_key_data else {
+            return Err(SessionKeyResolutionError { reason: "instance missing owner key/group data".to_string() })
+        };
+
+        let group_key = self.key_loader_facade.get_group_key(owner_group, owner_key_version)?;
+        Ok(group_key.decrypt_key(owner_enc_session_key).map(|k| Some(k))?)
+    }
+
+    fn resolve_bucket_key(&self, entity: &mut ParsedEntity, model: &TypeModel) -> Result<GenericAesKey, SessionKeyResolutionError> {
+        let Some(ElementValue::Dict(bucket_key_map)) = entity.get(BUCKET_KEY_NAME) else {
+            return Err(SessionKeyResolutionError { reason: format!("{BUCKET_KEY_NAME} is not a dictionary type") })
+        };
+
+        let bucket_key: BucketKey = match self.instance_mapper.parse_entity(bucket_key_map.to_owned()) {
+            Ok(n) => n,
+            Err(e) => return Err(SessionKeyResolutionError { reason: format!("{BUCKET_KEY_NAME} could not be deserialized: {e}") })
+        };
+
+        let owner_key_data = EntityOwnerKeyData::extract_owner_key_data(entity)?;
+        let Some(owner_group) = owner_key_data.owner_group else {
+            return Err(SessionKeyResolutionError { reason: "entity has no ownerGroup".to_owned() });
+        };
+
+        let VersionedAesKey { key, version } = self.key_loader_facade.get_current_group_key(owner_group)?;
+
+        let ResolvedBucketKey {
+            decrypted_bucket_key,
+            sender_identity_key
+        } = self.decrypt_bucket_key(&bucket_key, entity, model)?;
+
+        let mut session_key_for_this_instance = None;
+        let mut re_encrypted_session_keys = Vec::with_capacity(bucket_key.bucketEncSessionKeys.len());
+
+        for instance_session_key in bucket_key.bucketEncSessionKeys {
+            let decrypted_session_key = decrypted_bucket_key.decrypt_key(&instance_session_key.symEncSessionKey)?;
+            let iv = random(self.randomizer_facade.as_ref(), |rng| Iv::generate(rng));
+            let re_encrypted_session_key = decrypted_bucket_key.encrypt_key(&decrypted_session_key, iv);
+
+            if &instance_session_key.instanceId == owner_key_data.instance_id {
+                session_key_for_this_instance = Some((decrypted_session_key.clone(), re_encrypted_session_key.clone()));
+            }
+            re_encrypted_session_keys.push((instance_session_key, re_encrypted_session_key));
+        }
+
+        let Some((session_key, sym_enc_session_key)) = session_key_for_this_instance else {
+            return Err(SessionKeyResolutionError { reason: "no session key found in bucket key for this instance".to_string() })
+        };
+
+        // TODO: authenticate
```

Comment:
Should we implement authenticating the bucket key before merging this?

Reply:
I think we said we won't do it for the first version